DS3: Add the Dark Souls 3 World into Archipelago

Marechal-L opened the pull request:
First implementation of Dark Souls 3 into Archipelago.

The Closed Beta testing phase has been pretty good so I would like to start the Public Beta testing phase in parallel of this pull-request.

alwaysintreble commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,281 @@
+        data = self.item_name_to_id[name]
+        return DarkSouls3Item(
+            name, ItemClassification.progression if name in key_items_list else
+            ItemClassification.useful, data, self.player)
```
This causes generation failures as you have more items than locations with the default settings and none are marked as junk so none can be discarded. You should have a minimal number of progression and useful items so other people in the multiworld can get items too. Items and location counts should also never be unequal.

Marechal-L replied:
Fixed. I now use 3 differents classifications (progression, useful and filler) depending on the item category.

alwaysintreble commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,281 @@
+        if self.world.randomize_weapons_level[self.player]:
+            # Randomize some weapons upgrades
+            for name in weapons_upgrade_5_table.keys():
+                if randint(0, 100) < 33:
```
`self.world.random` object is seeded per generation and is deterministic. please use it instead

Marechal-L replied:
Fixed

alwaysintreble commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,281 @@
+    # For each region, add the associated locations retrieved from the corresponding location_table
+    def create_region(self, name, location_table) -> Region:
+        new_region = Region(name, RegionType.Generic, name, self.player)
+        if location_table is not None:
```
`if location_table:` is fine

Marechal-L replied:
Fixed

alwaysintreble commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,281 @@
+    def create_region(self, name, location_table) -> Region:
+        new_region = Region(name, RegionType.Generic, name, self.player)
+        if location_table is not None:
+            for name, address in self.location_name_to_id.items():
```
why not do this the other way around and just iterate through the supplied location_table instead of looping through the entirety of all locations for each region?

alwaysintreble commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,281 @@
+        else:
+            self.world.itempool += [item]
+
+        # Fill item pool with additional items
```
This doesn't do anything if the item pool is already larger than the location pool which it apparently is with the default settings.

Marechal-L replied:
I removed this part of the code as their is no more option to enter in this loop

alwaysintreble reviewed (commented):
few comments of things that I'm not sure I understand the intent of and some things that I believe don't work quite as intended. The purpose of extending the base Item and Location classes is if you're actually adding additional data to them. It's good practice to at least add your game name to your items and locations but isn't required and since you aren't doing that currently those classes aren't being used anyways. Pointed out a lot of the big ones. Should probably go through the item pool and decide which ones are actually necessary and useful and then mark the rest as junk. Was failing to generate for me and it doesn't help that even your filler item that gets created if your item pool < location pool is marked as useful. If a player marks any locations within their game as excluded (which any person can do at any time in any game in AP) this will also completely fail to generate regardless of other settings.

Marechal-L commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,281 @@
+    # The following two dicts are required for the generation to know which
+    # items exist. They could be generated from json or something else. They can
+    # include events, but don't have to since events will be placed manually.
+    item_name_to_id = {name: id for id, name in enumerate(DarkSouls3Item.get_item_name_to_id(), base_id)}
```
You're right, I created those classes before realising I didn't have custom data to add and I ended up using them as static classes with no real meaning.
Custom data could be introduced later but for now I removed them

Marechal-L commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,281 @@
+
+        data = {
+            "options": {
+                "auto_equip": (True if self.world.auto_equip[self.player] else False),
```
Fixed

Marechal-L commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,281 @@
+    # data_version is used to signal that items, locations or their names
+    # changed. Set this to 0 during development so other games' clients do not
+    # cache any texts, then increase by 1 for each release that makes changes.
+    data_version = 0
```
Fixed

Marechal-L commented:
Each location are now labeled as "Unknown location (ID:2030000)" with an incorrect ID.
I mark my pull-request as draft until I resolve this

black-sliver commented on worlds/dark_souls_3/docs/setup_en.md:
```diff
@@ -0,0 +1,35 @@
+The Dark Souls 3 AP Client is a dinput8.dll triggered when launching Dark Souls 3. This .dll file will launch a command 
+prompt where you can read information about your run and write any command to interact with the Archipelago server.  
+
+The randomization is performed by the AP.json file, an output file generated by the Archipelago server.
```
While this is probably fine in general, two possible ideas to make it easier to get started as a player:
* use slot_data that you get on connect instead of an external json file. - slot data is transferred as json, but you will only get it once connected and slot name is entered, so this may or may not work with the way you hook into the game - or
* use a different file extension (.apds3) and supply a program that will automatically put it in the correct spot with the correct name

Marechal-L replied:
Using the slot_data is a great idea! But as you said I will need to try if I would be able to hook the game at runtime and not at startup.
I'm starting to be running out of time for this project so I would probably only be able to provide this as an update :-/

black-sliver replied:
That's ok, it does not have to be perfect in the first iteration. Just wanted to let you know that those options exist.

black-sliver commented on worlds/dark_souls_3/docs/setup_en.md:
```diff
@@ -0,0 +1,35 @@
+## Joining a MultiWorld Game
+
+1. Run DarkSoulsIII.exe or run the game through Steam
+2. Type in /connect {SERVER_IP}:{SERVER_PORT}
```
where do you enter it? If it's a separate window (which i think it is), or if the player has to open a console, please say so.

Marechal-L replied:
A separate "Windows Command Prompt" is attached to the DarkSoulsIII.exe process at startup and launched automatically.
It may not be very convenient if the player only has one screen but it works pretty well.

black-sliver replied:
Would you mind putting that into the readme? Something like
> Type /connect ... in the "Windows Command Prompt" that opened

so that there is less room for error

black-sliver commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,280 @@
+        else:
+            item_classification = ItemClassification.filler
+
+        return Item(
+            name, item_classification, data, self.player)
```
That line break seems unintentional?

Marechal-L replied:
Fixed

black-sliver commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,280 @@
+    def set_rules(self) -> None:
+
+        # Define the access rules to the entrances
+        set_rule(
+            self.world.get_entrance("Goto Bell Tower", self.player),
+            lambda state: state.has("Mortician's Ashes", self.player))
```
In core code we format calls>120chars like so:
```python
        set_rule(self.world.get_entrance("Goto Bell Tower", self.player),
                 lambda state: state.has("Mortician's Ashes", self.player))
```
I don't have a strong opinion though, so if you think your version is better readable, I don't mind, and simply mark this as resolved.

Marechal-L replied:
Let's favor code consistency! It will make it more readable for all of us.
Fixed

black-sliver commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,280 @@
+        with open(os.path.join(output_directory, filename), 'w') as outfile:
+            json.dump(data, outfile)
+
+    game: str = "Dark Souls III"  # name of the game/world
+    options = dark_souls_options  # options the player can set
+    topology_present: bool = True  # show path to required location checks in spoiler
+    remote_items: bool = False  # True if all items come from the server
+    remote_start_inventory: bool = False  # True if start inventory comes from the server
+
+    # data_version is used to signal that items, locations or their names
+    # changed. Set this to 0 during development so other games' clients do not
+    # cache any texts, then increase by 1 for each release that makes changes.
+    data_version = 1
+
+    # ID of first item and location, could be hard-coded but code may be easier
+    # to read with this as a property.
+    base_id = 100000
+
+    # The following two dicts are required for the generation to know which
+    # items exist. They could be generated from json or something else. They can
+    # include events, but don't have to since events will be placed manually.
+    item_name_to_id = {name: id for id, name in enumerate(item_dictionary_table, base_id)}
+    location_name_to_id = {name: id for id, name in enumerate(location_dictionary_table, base_id)}
```
Please put the static class members at the beginning of your class

Marechal-L replied:
Fixed

black-sliver commented on worlds/dark_souls_3/Options.py:
```diff
@@ -0,0 +1,41 @@
+    display_name = "Late Basin of Vows"
+
+
+dark_souls_options: typing.Dict[str, type(Option)] = {
+    "auto_equip": AutoEquipOption,
+    "lock_equip": LockEquipOption,
+    "no_weapon_requirements": NoWeaponRequirementsOption,
+    "randomize_weapons_level": RandomizeWeaponsLevelOption,
+    "late_basin_of_vows": LateBasinOfVowsOption,
+}
```
This clearly misses death_link :-D

Marechal-L replied:
Ahah  yes, definitly!
I wanted to release this version and make it public first before adding some new features :D

black-sliver reviewed (commented):
Left a few comments. I did not test rolling or playing, but the code looks mostly sound.

I feel like we may want a `player=` filter in `MultiWorld.get_items` to make the generate_output cleaner.

Marechal-L commented:
Thanks @alwaysintreble and @black-sliver for your comments. I hope I addressed everything :-D

@black-sliver I may did not understand your remark about the "player= filter in MultiWorld.get_items".
I managed to remove one conditionnal check by using `self.world.get_filled_locations(self.player):` but your remark seems to be more destinated to the Core implementation, right ?

black-sliver commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,268 @@
+    remote_start_inventory: bool = False  # True if start inventory comes from the server
+    web = DarkSouls3Web()
+
+    # data_version is used to signal that items, locations or their names
+    # changed. Set this to 0 during development so other games' clients do not
+    # cache any texts, then increase by 1 for each release that makes changes.
```
Those comments can go. They are in the template for explanation, no need to replicate that in worlds.

black-sliver commented:
Oh, i screwed that second run of comments up big time xD
The first two are supposed to be replies to the two open discussions above.

> I may did not understand your remark about the "player= filter in MultiWorld.get_items".
I managed to remove one conditionnal check by using self.world.get_filled_locations(self.player): but your remark seems to be more destinated to the Core implementation, right ?

Yeah, I was talking about core. Your new code however produces different output than the old code. The old code added "All items from all worlds that are for your player" to the list, while the new code only adds "All items from your world that are for your player". Which one is correct? If it's the old one, you should probably revert. If it's the new one, the two `if`s can be merged.

Marechal-L commented:
I reverted my generate_ouput modification, I need to output all items from all worlds that are for the player because of the items_handling configuration.
I also cleaned some comments and updated the documentation as suggested :-D

Marechal-L commented:
Yes sure I wanted to schedule a public beta test this weekend :D
The steps are the following right ? 
- Schedule with AginahBot
- Collect all players .yaml
- Generate the game and share output AP_{SEED}.zip
- Host the AP server on my computer and share my public IP 
- Enjoy!

Berserker66 commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,258 @@
+        locations_address = []
+        locations_target = []
+        for location in self.world.get_filled_locations():
+            if location.item.player == self.player:
```
This makes me think this will break with item_links

Marechal-L replied:
I just set the item_links as follow for 2 players : 
```
item_links:
   - name: goMode
     item_pool: 
      - Cinders of a Lord - Abyss Watcher
      - Cinders of a Lord - Yhorm the Giant
      - Cinders of a Lord - Aldrich
      - Cinders of a Lord - Lothric Prince
     replacement_item: null 
```

And the nominal case seems to work as expected as the two players received the item : 
```
Server :         (Team #1) MarechToo sent Cinders of a Lord - Abyss Watcher to goMode (HWL: Longbow
Marech's client :   "#0: Cinders of a Lord - Abyss Watcher from MarechToo - HWL: Longbow\n",
MarechToo's client : "#0: Cinders of a Lord - Abyss Watcher from MarechToo - HWL: Longbow\n"
```

I would need help to list all different cases possible and an entire run to test them all.

Marechal-L commented on worlds/dark_souls_3/__init__.py:
```diff
@@ -0,0 +1,258 @@
+        else:
+            item_classification = ItemClassification.filler
+
+        return Item(name, item_classification, data, self.player)
```
Ok, so I should at least define a custom DarkSouls3Item class to assign the items to my game ?
I also have an idea to make it useful to re-write "ItemClassification" and "randomize_weapons_level" in a better way but I do not feel really confident in Python to do it as a first version.

Marechal-L replied:
Another question is, does it work the same for Locations ? Should I define a custom location class to at least assign them to my game ?

black-sliver replied:
We have the same thing for locations, yes.
Here is an example of the simplest version of those two: https://github.com/ArchipelagoMW/Archipelago/blob/e1ef820184c8ebd6d1f67cb3ec01d9e5764036f8/worlds/subnautica/__init__.py#L154-L159

Marechal-L replied:
Done

Marechal-L commented on worlds/dark_souls_3/docs/setup_en.md:
```diff
@@ -0,0 +1,35 @@
+# Dark Souls 3 Randomizer Setup Guide
```
I was in struggle with myself between "III" and "3" but you're right "III" looks definitely better in documentations and texts.

Marechal-L replied:
Done

Marechal-L commented on worlds/dark_souls_3/docs/en_Dark Souls III.md:
```diff
@@ -0,0 +1,22 @@
+# Dark Souls 3
```
Done. And even more as I renamed my AP Client github project
